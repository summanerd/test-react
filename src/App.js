import { useCallback, useState } from 'react';
import './App.css';

const initialState = {
  firstName: '',
  email: ''
};

const isValid = ({ firstName, email }) => !!(firstName && email) && firstName.length > 2 && email.match(/@\w+\.com$/);

function App({onSubmit}) {
  const [formValues, setFormValues] = useState(initialState);
  const isSubmitDisabled = !isValid(formValues);
  const onChange = useCallback((e, prop)=> setFormValues((prev)=>({...prev, [prop]: e.target.value})), [setFormValues])

  return (
    <div className="App">
      <header className="App-header">
        <h2>Test React</h2>
      </header>
      <main>
        <form data-testid="form" onSubmit={()=>onSubmit(formValues)}>
          <input type="text"
            name="firstName"
            value={formValues.firstName}
            onChange={event => onChange(event, 'firstName')}
            data-testid="firstName" />
          <input type="email"
            name="email"
            value={formValues.email}
            onChange={event => onChange(event, 'email')}
            data-testid="email" />
          <button type="submit" disabled={isSubmitDisabled} >Submit</button>
        </form>
      </main>
    </div>
  );
}

export default App;
