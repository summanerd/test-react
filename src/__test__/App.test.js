import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import App from '../App';

const user = userEvent.setup()

test('initial render', () => {
  const { container } = render(<App onSubmit={jest.fn()} />);
  expect(container).toMatchSnapshot();
  expect(screen.getByRole('button')).toHaveAttribute('disabled')
});

test('partially completed', async () => {
  const { container } = render(<App onSubmit={jest.fn()} />);
  const nameInput = screen.getByTestId('firstName');

  await user.tab()
  expect(nameInput).toHaveFocus()
  await user.keyboard('mauricia');

  expect(container).toMatchSnapshot();
  expect(screen.getByRole('button')).toHaveAttribute('disabled')
});

test('invalid email', async () => {
  const { container } = render(<App onSubmit={jest.fn()} />);
  const nameInput = screen.getByTestId('firstName');
  const emailInput = screen.getByTestId('email');
  
  await user.tab()
  expect(nameInput).toHaveFocus()
  await user.keyboard('mauricia');
  
  await user.tab()
  expect(emailInput).toHaveFocus()
  await user.keyboard('mauricia@');

  expect(container).toMatchSnapshot();
  expect(screen.getByRole('button')).toHaveAttribute('disabled')
});

test('valid firstName and email', async () => {
  const { container } = render(<App onSubmit={jest.fn()} />);
  const nameInput = screen.getByTestId('firstName');
  const emailInput = screen.getByTestId('email');
  
  await user.tab()
  expect(nameInput).toHaveFocus()
  await user.keyboard('mauricia');
  
  await user.tab()
  expect(emailInput).toHaveFocus()
  await user.keyboard('mauricia@summanerd.com');

  expect(container).toMatchSnapshot();
  expect(screen.getByRole('button')).not.toHaveAttribute('disabled')
});

test('submit valid form', async () => {
  const onSubmit = jest.fn();
  render(<App onSubmit={onSubmit} />);
  const nameInput = screen.getByTestId('firstName');
  const emailInput = screen.getByTestId('email');
  const form = screen.getByTestId('form');
  
  await user.tab()
  expect(nameInput).toHaveFocus()
  await user.keyboard('mauricia');
  
  await user.tab()
  expect(emailInput).toHaveFocus()
  await user.keyboard('mauricia@summanerd.com');

  await fireEvent.submit(form);

  expect(onSubmit).toHaveBeenCalled();
  const [[args]] = onSubmit.mock.calls
  expect(args).toEqual({firstName: 'mauricia', email: 'mauricia@summanerd.com'});
});
