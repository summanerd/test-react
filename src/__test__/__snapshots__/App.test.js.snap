// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial render 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h2>
        Test React
      </h2>
    </header>
    <main>
      <form
        data-testid="form"
      >
        <input
          data-testid="firstName"
          name="firstName"
          type="text"
          value=""
        />
        <input
          data-testid="email"
          name="email"
          type="email"
          value=""
        />
        <button
          disabled=""
          type="submit"
        >
          Submit
        </button>
      </form>
    </main>
  </div>
</div>
`;

exports[`invalid email 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h2>
        Test React
      </h2>
    </header>
    <main>
      <form
        data-testid="form"
      >
        <input
          data-testid="firstName"
          name="firstName"
          type="text"
          value="mauricia"
        />
        <input
          data-testid="email"
          name="email"
          type="email"
          value="mauricia@"
        />
        <button
          disabled=""
          type="submit"
        >
          Submit
        </button>
      </form>
    </main>
  </div>
</div>
`;

exports[`partially completed 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h2>
        Test React
      </h2>
    </header>
    <main>
      <form
        data-testid="form"
      >
        <input
          data-testid="firstName"
          name="firstName"
          type="text"
          value="mauricia"
        />
        <input
          data-testid="email"
          name="email"
          type="email"
          value=""
        />
        <button
          disabled=""
          type="submit"
        >
          Submit
        </button>
      </form>
    </main>
  </div>
</div>
`;

exports[`submit valid form 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h2>
        Test React
      </h2>
    </header>
    <main>
      <form>
        <input
          data-testid="firstName"
          name="firstName"
          type="text"
          value="mauricia"
        />
        <input
          data-testid="email"
          name="email"
          type="email"
          value="mauricia@summanerd.com"
        />
        <button
          data-testid="submit"
          type="submit"
        >
          Submit
        </button>
      </form>
    </main>
  </div>
</div>
`;

exports[`valid firstName and email 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h2>
        Test React
      </h2>
    </header>
    <main>
      <form
        data-testid="form"
      >
        <input
          data-testid="firstName"
          name="firstName"
          type="text"
          value="mauricia"
        />
        <input
          data-testid="email"
          name="email"
          type="email"
          value="mauricia@summanerd.com"
        />
        <button
          type="submit"
        >
          Submit
        </button>
      </form>
    </main>
  </div>
</div>
`;
